### PR TITLE
[Behavorial Analytics] CRUD REST APIs 

### PR DESCRIPTION
**Pull request by joemcelroy**
PR contains:
- PUT, GET & DELETE REST actions for analytics collections.
- Introduces a `manage_behavioral_analytics` cluster privilege.
- Each action uses the new privilege. 

### PUT

```
PUT /_behavioral_analytics/test2!
{} 
```

Response:
```json
{
    "test2!": {
        "event_data_stream": {
            "name": "behavioral_analytics-events-test2!"
        }
    }
}
```

### GET

```
GET /_behavioral_analytics/test2!
```

Response:
```json
{
    "test2!": {
        "event_data_stream": {
            "name": "behavioral_analytics-events-test2!"
        }
    }
}
```

### analytics collection not found exception

```
GET /_behavioral_analytics/dont_exist
```

Response:
```json
{
    "error": {
        "root_cause": [
            {
                "type": "resource_not_found_exception",
                "reason": "test2!n"
            }
        ],
        "type": "resource_not_found_exception",
        "reason": "test2!n"
    },
    "status": 404
}
```

### Delete
```
DELETE _behavioral_analytics/test2!
```

Response:
```json
{
    "acknowledged": true
}
```

### CODE REVIEW

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/AnalyticsCollection.java [hunk 1]
```diff
@@ -79,6 +79,8 @@ public String getEventDataStream() {
     @Override
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject();
+        builder.field("name", this.getName());
```
Comment (joemcelroy) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/AnalyticsCollection.java:82]: should these be constants?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/TransportGetAnalyticsCollectionAction.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    protected void masterOperation(Task task, GetAnalyticsCollectionAction.Request request, ClusterState state, ActionListener<GetAnalyticsCollectionAction.Response> listener) {
+        analyticsCollectionService.getAnalyticsCollection(
+            request.getCollectionName(),
+            listener.map(GetAnalyticsCollectionAction.Response::new)
+        );
+    }
+
+    @Override
+    protected ClusterBlockException checkBlock(GetAnalyticsCollectionAction.Request request, ClusterState state) {
+        return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_READ);
```
Comment (joemcelroy) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/TransportGetAnalyticsCollectionAction.java:62]: Aurelien - this correct using Metadata_read?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/PutAnalyticsCollectionAction.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+            analyticsCollection = new AnalyticsCollection(in);
+        }
+
+        public Response(AnalyticsCollection analyticsCollection) {
+            this.analyticsCollection = analyticsCollection;
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject();
```
Comment (joemcelroy) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/PutAnalyticsCollectionAction.java:104]: advice needed where I could move this to standardize the response.

Comment (joemcelroy) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/PutAnalyticsCollectionAction.java]: for now i've introduced a BaseAnalyticsCollectionResponse class to handle this

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/AnalyticsCollection.java (hunk 1 above)
Comment (joemcelroy) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/AnalyticsCollection.java]: moved to using parser_field

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/ClearDeploymentCacheAction.java [hunk 4]
```diff
@@ -98,5 +98,7 @@ public XContentBuilder toXContent(XContentBuilder builder, ToXContent.Params par
             builder.endObject();
             return builder;
         }
+
```
Comment (joemcelroy) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/ClearDeploymentCacheAction.java:101]: not sure why this change was included. Looking into it

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/PutAnalyticsCollectionAction.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+public class PutAnalyticsCollectionAction extends ActionType<PutAnalyticsCollectionAction.Response> {
+
+    public static final PutAnalyticsCollectionAction INSTANCE = new PutAnalyticsCollectionAction();
+    public static final String NAME = "cluster:admin/behavioral_analytics/put";
+
+    public PutAnalyticsCollectionAction() {
+        super(NAME, PutAnalyticsCollectionAction.Response::new);
+    }
+
+    public static class Request extends MasterNodeRequest<Request> {
```
Comment (jimczi) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/PutAnalyticsCollectionAction.java:38]: Same here, why MasterNodeRequest?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/TransportDeleteAnalyticsCollectionAction.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+import org.elasticsearch.cluster.block.ClusterBlockLevel;
+import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.entsearch.analytics.AnalyticsCollectionService;
+
+public class TransportDeleteAnalyticsCollectionAction extends AcknowledgedTransportMasterNodeAction<
```
Comment (jimczi) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/TransportDeleteAnalyticsCollectionAction.java:25]: A `HandledTransportAction` should be enough, there's no need to make it a master node only action?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/TransportPutAnalyticsCollectionAction.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.entsearch.analytics.AnalyticsCollection;
+import org.elasticsearch.xpack.entsearch.analytics.AnalyticsCollectionService;
+
+public class TransportPutAnalyticsCollectionAction extends TransportMasterNodeAction<
```
Comment (jimczi) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/TransportPutAnalyticsCollectionAction.java:24]: Same here

**Review by jimczi — COMMENTED** [review]
The logic seems good to me, however I'd like to understand why these actions need to be master node actions. That doesn't seem right.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/PutAnalyticsCollectionAction.java (hunk 5 above)
Comment (joemcelroy) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/PutAnalyticsCollectionAction.java:38]: Leaning on @afoucret  to give a better answer here but we are using the MasterNodeRequest due to the way we use `clusterChanged` event to update the cached list of analytics collections in the `AnalyticsCollectionService`. From what I understand, we need to wait for master node acknowledgement so that the cluster metadata event has been already emitted and cache updated before we respond back.  

Happy to use the normal `ActionRequest` if you don't think this is right.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/GetAnalyticsCollectionAction.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+        public Request(String collectionName) {
+            this.collectionName = collectionName;
+        }
+
+        @Override
+        public ActionRequestValidationException validate() {
+            ActionRequestValidationException validationException = null;
+
+            if (collectionName == null || collectionName.isEmpty()) {
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/GetAnalyticsCollectionAction.java:51]:
```suggestion
            if (Strings.isNullOrEmpty(collectionName)) {
```

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/PutAnalyticsCollectionAction.java [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+            return Objects.equals(analyticsCollection, that.analyticsCollection);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(analyticsCollection);
+        }
+    }
+
+    public static class Response extends BaseAnalyticsCollectionResponse implements ToXContentObject {
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/PutAnalyticsCollectionAction.java:91]: For Engines, the  PUT API is returning just "result": "created" or "updated". This is probably more similar to what other ES APIs do today.

Do you have a use case for returning the whole analytics collection information?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/support/BaseAnalyticsCollectionResponse.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.ToXContent;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xpack.entsearch.analytics.AnalyticsCollection;
+
+import java.io.IOException;
+import java.util.Objects;
+
+public class BaseAnalyticsCollectionResponse extends ActionResponse {
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/support/BaseAnalyticsCollectionResponse.java:21]: Good call capturing common code in this class! If PUT action does not need to return analytics information, then it might be unnecessary.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/PutAnalyticsCollectionAction.java (hunk 5 above)
Comment (jimczi) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/PutAnalyticsCollectionAction.java:38]: Ok I see, then I wonder if we should change the AnalyticsCollectionService to be available on the master node only? @afoucret are we expecting the service to be available on every node?

**Review by carlosdelest — COMMENTED** [review]
Some minor questions. LGTM otherwise 💯

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/support/BaseAnalyticsCollectionResponse.java (hunk 10 above)
Comment (joemcelroy) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/support/BaseAnalyticsCollectionResponse.java:21]: Yeh its an interesting point. Right now i've kept it to share the same response model as previous API did so. I think the base response class will come useful more when LIST is implemented too. Also potentially in future an analytics collection may be made up from a couple of service calls (when we get into sessions and Users).

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/PutAnalyticsCollectionAction.java (hunk 9 above)
Comment (joemcelroy) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/PutAnalyticsCollectionAction.java:91]: ATM the design has been like for like on the previous API. We were originally required the whole model back for updates so we refresh the client model without needing to make another call to GET endpoint. That update mechanism will likely change now we are using datastreams directly. 

Will leave it for now as model but possible change in contract when we get into kibana integration detail.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/EnterpriseSearch.java [hunk 11]
```diff
@@ -63,6 +72,7 @@
 
 public class EnterpriseSearch extends Plugin implements ActionPlugin, SystemIndexPlugin {
     public static final String ENGINE_API_ENDPOINT = "_engine";
+    public static final String BEHAVIORAL_ANALYTICS_API_ENDPOINT = "_behavioral_analytics";
```
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/EnterpriseSearch.java:75]: Just curious, did you consider `_analytics` or something similar, as it would be a shorter endpoint?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/DeleteAnalyticsCollectionAction.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+        public String getCollectionName() {
+            return collectionName;
+        }
+
+        @Override
+        public ActionRequestValidationException validate() {
+            ActionRequestValidationException validationException = null;
+
+            if (this.collectionName == null || this.collectionName.isEmpty()) {
```
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/DeleteAnalyticsCollectionAction.java:51]: Nitpick: `Strings.isNullOrEmpty(collectionName)`

**Review by kderusso — COMMENTED** [review]
Looking good so far

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/EnterpriseSearch.java (hunk 11 above)
Comment (joemcelroy) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/EnterpriseSearch.java:75]: The challenge is the url isn't part of the engine's path so  `HOST/_analytics` might be too generic. We also have an analytics module in elasticsearch so want the name to be more targeted to us.  I think `_analytics` makes sense if it was `HOST/_engine/<engine>/analytics`

**Review by jimczi — APPROVED** [review]
LGTM, thanks @joemcelroy . Let's add the license check in a follow up.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/PutAnalyticsCollectionAction.java (hunk 5 above)
Comment (joemcelroy) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/PutAnalyticsCollectionAction.java:38]: I believe it needs to be available on every node as we require the analyticsCollections cache to be used by the event intake API. We need to verify that the analytics collection exists before we can emit the event to the events log.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/PutAnalyticsCollectionAction.java [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+            super.writeTo(out);
+            analyticsCollection.writeTo(out);
+        }
+
+        public AnalyticsCollection getAnalyticsCollection() {
+            return analyticsCollection;
+        }
+
+        @Override
+        public boolean equals(Object o) {
```
Comment (ioanatia) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/PutAnalyticsCollectionAction.java:78]: there are a few methods here that are duplicated in the other `Request` classes - could we potentially try to extract them into a different class? not necessarily to be done here, we could look into this in another PR.

Comment (joemcelroy) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/analytics/action/PutAnalyticsCollectionAction.java:78]: Yeh I think so - theres definitely room for reducing boilerplate, similar to the BaseAnalyticsCollectionRequest. I will have a look what other solutions do for this and keep it in mind!
